fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,211 @@
+use fvm_shared::clock::ChainEpoch;
+use ipc_agent_sdk::apis::IpcAgentClient;
+use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::subnet_id::SubnetID;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer {
+    config: Config,
+    parent_view_provider: Arc<InMemoryFinalityProvider>,
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl PollingParentSyncer {
+    pub fn new(
+        config: Config,
+        parent_view_provider: Arc<InMemoryFinalityProvider>,
+        agent: Arc<IPCAgentProxy>,
+    ) -> Self {
+        Self {
+            config,
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl PollingParentSyncer {
+    /// Start the parent finality listener in the background
+    pub fn start(self) -> anyhow::Result<()> {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+
+        tokio::spawn(async move {
+            loop {
+                // Syncing with parent with the below steps:
+                // 1. Get the latest height in cache or latest height committed increment by 1 as the
+                //    starting height
+                // 2. Get the latest chain head height deduct away N blocks as the ending height
+                // 3. Fetches the data between starting and ending height
+                // 4. Update the data into cache
+                if let Err(e) = sync_with_parent(&config, &agent, &provider).await {
+                    tracing::error!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCAgentProxy>,
+    provider: &Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+    loop {
+        interval.tick().await;
+
+        let starting_height = get_starting_height(provider).await?;
+        let latest_height = agent_proxy
+            .get_chain_head_height()
+            .await
+            .context("cannot fetch parent chain head")?;
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+        let ending_height = latest_height - config.chain_head_delay;
+
+        tracing::debug!("starting height: {starting_height}, ending_height: {ending_height}");
+
+        // we are going backwards in terms of block height, the latest block height is lower
+        // than our previously fetched head. It could be a chain reorg. We clear all the cache
+        // in `provider` and start from scratch
+        if starting_height > ending_height {
+            todo!()
+        }
```

Comment:
RIght, I removed the `todo` and made the checks more explicit:
```rust
        let last_recorded_height = last_recorded_height(provider).await?;
        let ending_height = ... // some logic omitted for the current discussion;

        tracing::debug!(
            "last recorded height: {}, parent chain head: {}, ending_height: {}",
            last_recorded_height, parent_chain_head_height, ending_height
        );

        if last_recorded_height == ending_height {
            tracing::debug!("the parent has yet to produce a new block, stops at height: {last_recorded_height}");
            continue;
        }

        // we are going backwards in terms of block height, the latest block height is lower
        // than our previously fetched head. It could be a chain reorg. We clear all the cache
        // in `provider` and start from scratch
        if last_recorded_height > ending_height {
            todo!()
        }
```
I still think we have to resync with parent when `last_recorded_height > ending_height` as we still rely on caching for `validating`. So in ` The syncer would consider 91 final now, and attempt to fetch from 93 to 91, hitting this condition and fail, while in fact this is all fine, as long as 92 isn't rolled back.`, even if `92` is all fine, but 91 might have invalid block hash and `process_proposal` might fail. Maybe we can just reset and restart caching then we should be ok.